File: Reference Bots/C#/ReferenceBot/Strategy/BasicShootStrategy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ReferenceBot.Domain.Command;
using ReferenceBot.Domain.Command.Ship;
using ReferenceBot.Domain.State;

namespace ReferenceBot.Strategy
{
    public class BasicShootStrategy
    {
        private const int smallMapSize = 7;
        private const int mediumMapSize = 10;
        private const int largeMapSize = 14;

        private int energyPerRound = 2;

        public Command ExecuteStrategy(GameState gameState)
        {
            return RandomShotCommand(gameState);
        }
        
        private Command RandomShotCommand(GameState gameState)
        {
            var cells = gameState.OpponentMap.Cells;
            var lastShot = cells.OrderBy(cell => cell.Y).ThenBy(cell => cell.X).LastOrDefault(cell => cell.Damaged || cell.Missed);
            if (lastShot == null)
            {
                //We have not made any shots yet, shot in the first coordinate
                return new Command(Code.FireShot, 0, 0);
            }

            
            var x = lastShot.X;
            var y = lastShot.Y;

            if (x + 2 < gameState.PlayerMap.MapWidth)
            {
                x += 2;
            }
            else
            {
                x = 0;
                y++;
            }

            if (y >= gameState.PlayerMap.MapWidth)
            {
                return AlternateRandomShot(gameState);
            }

            switch (gameState.MapDimension)
            {
                case mediumMapSize:
                    energyPerRound = 3;
                    break;
                case largeMapSize:
                    energyPerRound = 4;
                    break;
            }

            var currentEnergy = gameState.PlayerMap.Owner.Energy;

            var destroyerAvailable =
                gameState.PlayerMap.Owner.Ships.FirstOrDefault(a => a.ShipType == ShipType.Destroyer && !a.Destroyed);

            var doubleShotWeaponEnergyRequired = destroyerAvailable?.Weapons
                .Single(c => c.WeaponType == WeaponType.DoubleShot).EnergyRequired;

            var shieldAvailable = !gameState.PlayerMap.Owner.Shield.Active &&
                                  gameState.PlayerMap.Owner.Shield.CurrentCharges >= 1;

            //This will place a shield over the destroyer if it is not destroyed
            if (destroyerAvailable != null && shieldAvailable)
            {
                var cell = destroyerAvailable.Cells[0];
                return new Command(Code.Shield, cell.X, cell.Y);
            }

            if (doubleShotWeaponEnergyRequired != null && currentEnergy >= doubleShotWeaponEnergyRequired)
            {
                return new Command(Code.DoubleShotVertical, x, y);
            }

            return new Command(Code.FireShot, x, y);
        }

        private Command AlternateRandomShot(GameState gameState)
        {
            var cells = gameState.OpponentMap.Cells;
            var availableCell = cells.FirstOrDefault(cell => !cell.Damaged && !cell.Missed);

            if (availableCell == null)
            {
                var random = new Random();
                return new Command(Code.FireShot, random.Next(0, gameState.PlayerMap.MapWidth), random.Next(gameState.PlayerMap.MapHeight));
            }
            return new Command(Code.FireShot, availableCell.X, availableCell.Y);
        }
    }
}

File: Reference Bots/Java/ReferenceBot/src/main/java/za/co/entelect/challenge/strategy/BasicShootStrategy.java
package za.co.entelect.challenge.strategy;

import za.co.entelect.challenge.domain.command.Command;
import za.co.entelect.challenge.domain.command.code.Code;
import za.co.entelect.challenge.domain.command.direction.Direction;
import za.co.entelect.challenge.domain.command.ship.ShipType;
import za.co.entelect.challenge.domain.state.GameState;
import za.co.entelect.challenge.domain.state.OpponentCell;
import za.co.entelect.challenge.domain.state.Ship;
import za.co.entelect.challenge.domain.state.WeaponType;

import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

public class BasicShootStrategy {

    public Command executeStrategy(GameState gameState) {
        return randomShotCommand(gameState);
    }

    private Command randomShotCommand(GameState gameState) {
        ArrayList<OpponentCell> opponentCells = gameState.OpponentMap.Cells;
        Optional<OpponentCell> lastShot = opponentCells.stream()
                .sorted(Comparator.comparingInt(y -> y.Y))
                .sorted(Comparator.comparingInt(y -> y.X))
                .filter(cell -> cell.Damaged || cell.Missed).findFirst();

        if (!lastShot.isPresent()) {
            return new Command(Code.FIRE_SHOT, 0, 0);
        }

        int x = lastShot.get().X;
        int y = lastShot.get().Y;

        if (x + 2 < gameState.PlayerMap.MapWidth) {
            x += 2;
        } else {
            x = 0;
            y++;
        }
        if (y >= gameState.PlayerMap.MapWidth) {
            return alternateRandomShot(gameState);
        }

        int currentEnergy = gameState.PlayerMap.Owner.Energy;

        Optional<Ship> destroyerAvailable = gameState.PlayerMap.Owner.Ships.stream().filter(z -> z.ShipType == ShipType.Destroyer && !z.Destroyed).findFirst();

        if(destroyerAvailable.isPresent()) {
            int energyRequired = destroyerAvailable.get().Weapons.stream().filter(z -> z.WeaponType == WeaponType.DoubleShot).findFirst().get().EnergyRequired;
            if(currentEnergy >= energyRequired) {
                return new Command(Code.DOUBLE_SHOT_HORIZONTAL ,x, y);
            }
        }

        return new Command(Code.FIRE_SHOT, x, y);
    }

    private Command alternateRandomShot(GameState gameState) {
        ArrayList<OpponentCell> opponentCells = gameState.OpponentMap.Cells;
        Optional<OpponentCell> availableCell = opponentCells.stream().filter(x -> !x.Damaged && !x.Missed).findFirst();

        if (!availableCell.isPresent()) {
            return new Command(Code.FIRE_SHOT, ThreadLocalRandom.current().nextInt(0, gameState.PlayerMap.MapWidth), ThreadLocalRandom.current().nextInt(gameState.PlayerMap.MapHeight));
        }
        return new Command(Code.FIRE_SHOT, availableCell.get().X, availableCell.get().Y);
    }
}
